math/vecops/inc/ROOT/RVec.hxx

```diff
@@ -94,32 +111,985 @@ auto MapFromTuple(Tuple_t &&t, std::index_sequence<Is...>)
    return MapImpl(std::get<tupleSizeM1>(t), std::get<Is>(t)...);
 }
 
+} // namespace VecOps
+} // namespace Detail
+
+namespace Internal {
+namespace VecOps {
+/// Return the next power of two (in 64-bits) that is strictly greater than A.
+/// Return zero on overflow.
+inline uint64_t NextPowerOf2(uint64_t A)
+{
+   A |= (A >> 1);
+   A |= (A >> 2);
+   A |= (A >> 4);
+   A |= (A >> 8);
+   A |= (A >> 16);
+   A |= (A >> 32);
+   return A + 1;
 }
+
+/// This is all the stuff common to all SmallVectors.
+///
+/// The template parameter specifies the type which should be used to hold the
+/// Size and Capacity of the SmallVector, so it can be adjusted.
+/// Using 32 bit size is desirable to shrink the size of the SmallVector.
+/// Using 64 bit size is desirable for cases like SmallVector<char>, where a
+/// 32 bit size would limit the vector to ~4GB. SmallVectors are used for
+/// buffering bitcode output - which can exceed 4GB.
+class SmallVectorBase {
+public:
+   using Size_T = int32_t;
+
+protected:
+   void *fBeginX;
+   /// Always >= 0.
+   // Type is signed only for consistency with fCapacity.
+   Size_T fSize = 0;
+   /// Always >= -1. fCapacity == -1 indicates the RVec is in "memory adoption" mode.
+   Size_T fCapacity;
```

Comment:
What is the reason for not initializing `fBeginX` and `fCapacity`, but `fSize`? It does not introduce overhead if you then initialize them differently in a ctor, but gives you safety in case you will add a ctor at some point.

Reply:
This comes from llvm's SmallVector implementation, so just guessing: on one hand there are no very good defaults here, it's _never_ ok  to have `fBeginX = nullptr`, for instance -- you always need the constructor to assign something sensible. And if you add a default that does not make sense it silences valgrind and static analyzers when you then forget to initialize to something sensible.

But now that you point it out I don't know what to do, how strongly do you feel about initializing these data members to values that most likely don't make sense? :smile:

Reply:
I prefer determinism over the chance to catch an incorrect use of `SmallVector`'s API in valgrind. If `fBeginX` always needs to be initialized, then there should be no constructor that leaves it uninitialized and so giving it a member initializer incurs no cost but gives you potential safety in case someone adds a new ctor.

Reply:
The LLVM folks argue exactly the inverse: according to them it's important to allow tooling to catch memory errors like uninitialized members. That's why [LLVM's code does not initialize these members](https://llvm.org/doxygen/SmallVector_8h_source.html#l00047).

Reply:
Well, the LLVM folks also have `-ftrivial-auto-var-init=zero` which just let's the compiler do it. Do you turn that on for release builds of ROOT used in production? And if you leave the members uninitialized, will your users run the tools to catch their memory errors?

Reply:
again there is no value that makes sense there, everything is broken if those data members are not initialized by the constructor. as there are conflicting opinions on this I prefer sticking to the original implementation.